Test loading of cached metadata in ngclient

ivanayov opened the pull request:
After making a successful update of valid metadata which stores it
in cache and performing a second update with a new updater while
the metadata is already stored in cache, this test verifies that
timestamp, snaphot and targets are loaded from cache and not 
downloaded

Addresses #1681

Signed-off-by: Ivana Atanasova <iyovcheva@vmware.com>

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [n/a] Docs have been added for the bug fix or new feature

Commit:
Test loading of cached metadata in ngclient

After making a successful update of valid metadata which stores it
in cache and performing a second update with a new updater while
the metadata is already stored in cache, this test verifies that
timestamp, snaphot and targets are loaded from cache and not
downloaded

Fixes #1681

Signed-off-by: Ivana Atanasova <iyovcheva@vmware.com>

kairoaraujo commented:
Hi @ivanayov, I think my PR #1704  will help you to remove the mock part. You will be able to test it directly against the Repository Simulator.

jku commented on tests/test_updater_with_simulator.py:
```diff
@@ -257,6 +257,42 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
+
+            # Test that snaphot and targets are loaded from cache and not downloaded
+            # and that timestamp is loaded from cache and used to verify the new timestamp
+            wrapped_fetch.assert_has_calls([call("root", 2), call("timestamp", None)])
```
testing this has some merit... but I bet we could add this test to `test_top_level_roles_update()`and `test_delegated_roles_update()`:  existing tests that already test whether the expected files are being fetched.

However, I would like to _actually verify what we load from disk_. The PR currently does not do that. So my (possibly naive) ideas was to patch the builtin open() and see what gets loaded from disk during `Updater()`, `refresh()` and `get_targetinfo()`

ivanayov commented:
@jku I addressed your comment, if something is not very correct I'll edit it. (It doesn't cover if the loaded timestamp is used to verify the new timestamp, but AFAIR it's already covered and we wanted to check only what's loaded from disk.)

Thanks @kairoaraujo . With the latest change I don't check `fetch` anymore, but will keep in mind for future tests.

MVrachev commented:
Could you please rebase on top of develop?
In #1710 we started linting our tests and it will be good to lint the changes made in this pr.

jku commented on tests/test_updater_with_simulator.py:
```diff
@@ -257,6 +257,44 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
+        # Clean up calls to fetch during refresh()
+        wrapped_open.reset_mock()
+
+        updater.get_targetinfo(self.targets_dir)
```
I think this call does not need to be here (and the argument is weird)

sechkova replied:
+1. This call only makes sense if a delegated role was added to the repository.

jku commented on tests/test_updater_with_simulator.py:
```diff
@@ -257,6 +257,44 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
+            "https://example.com/targets/",
+            self.sim,
+        )
+        updater.get_targetinfo(self.targets_dir)
```
Calling `get_targetinfo()` feels correct but the argument is confusing: maybe just use `"non_existent_target"` or something like that as target_path -- the value does not matter but using `targets_dir` is misleading.

jku reviewed (changes requested):
Yes, this looks ok. Left some comments on get_targetinfo() calls: please fix those.

Since you are already modifying I'll leave another nitpick: can you double check that comments and commit messages do not claim to do something that isn't done: at least one comment says _"Test that timestamp, snaphot and targets are ... not downloaded"_ which is not true. The test only checks that local files are opened in the expected way (and that's fine, this is useful and would have caught a past issue we had).

There is a potential follow-up for 
* adding a delegated target to repository (se the test can check that delegated target is also correctly loaded from local disk)
* separately testing constructor, refresh() and get_targetinfo()

sechkova commented on tests/test_updater_with_simulator.py:
```diff
@@ -257,6 +257,44 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
         wrapped_open.assert_not_called()
 
+    @patch.object(builtins, "open", wraps=builtins.open)
+    def test_update_metadata_to_cache(self, wrapped_open: MagicMock) -> None:
```
`test_load_metadata_from_cache` or something like it would match better what this test is trying to achieve.

sechkova commented on tests/test_updater_with_simulator.py:
```diff
@@ -257,6 +257,44 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
+        # Add targets to repository
+        self.sim.targets.version += 1
+        self.sim.add_target("targets", b'some content', self.targets_dir)
+        self.sim.update_snapshot()
```
Adding target files is not needed at all for this test. Target files, at least in this case, are unrelated to metadata files. Adding a delegated role is what would be useful so that delegated metadata is tested too.

sechkova commented on tests/test_updater_with_simulator.py:
```diff
@@ -257,6 +257,44 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
+            call(os.path.join(self.metadata_dir, "snapshot.json"), "rb"),
+            call(os.path.join(self.metadata_dir, "targets.json"), "rb"),
+        ])
+        wrapped_open.reset_mock()
```
The test is executed only once, this reset causes no harm but is also useless.

sechkova reviewed (changes requested):
I added some comments to the code but then realized, this test does not prove much. Updater always tries to load metadata from cache. There will always be calls to open even if there are no cached files. If this is what we want to demonstrate, ok. Then we are testing that "Updater always tries to read the cache before downloading metadata" and a single call to "refresh" is enough. The whole test would be:
```
updater = self._run_refresh()
wrapped_open.assert_has_calls([
            call(os.path.join(self.metadata_dir, "root.json"), "rb"),
            call(os.path.join(self.metadata_dir, "timestamp.json"), "rb"),
            call(os.path.join(self.metadata_dir, "snapshot.json"), "rb"),
            call(os.path.join(self.metadata_dir, "targets.json"), "rb"),
        ])
```

Probably a way to show that it opened and used the cached metadata is to show that when cached metadata is loaded from the file system successfully, there are no calls to RepositorySimulator.fetch_metadata for snapshot and targets. This should be easier now, after the addition of RepositorySimulator.fetch_counters. "Usage" of local (intermediate) root is described in the spec and should have be covered in test_top_level_update.py. Timestamp is a special case that can be tested in another set of tests covering invalid metadata.

The test is not really consistent. It randomly calls `_run_refresh` and `get_targetinfo`.
In case it is not clear, the options are either:
1. Test only top-level roles metadata 
 - create a new updater, call updater.refresh to get the metadata locally
 - reset the mock
 - create a new updater, call updater.refresh a second time to get the metadata loaded from the local cache dir
 - check the calls to open, check the calls to RepositorySimulator.fetch_tracker.metadata

Then as Jussi suggests, add delegations as follow up.

2. Test both top-level metadata and a delegated role
- add a delegated role to simulator, there is an example in https://github.com/theupdateframework/python-tuf/blob/bdf1cbbbb56839cd99b965fedae2661ed873c4a2/tests/test_updater_with_simulator.py#L148
- create a new updater, call updater.get_targetinfo("non_existent_target") to get the metadata locally, delegated role included
- reset the mock
 - create a new updater, call updater.get_targetinfo("non_existent_target")  a second time to get the metadata loaded from the local cache dir
 - check the the calls to open, check the calls to RepositorySimulator.fetch_tracker.metadata

sechkova commented on tests/test_updater_with_simulator.py:
```diff
@@ -254,6 +254,44 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
+
+        # Make a successful update of valid metadata which stores it in cache
+        updater = self._run_refresh()
+        updater.get_targetinfo("non_existent_target")
```
If you look at get_targetinfo, you'll see it calls refresh, so no need for a double refresh here, only creating a new updater instance + updater.get_targetinfo.

jku replied:
on the other hand, creating a new updater is multiple lines of code because there's no helper function for that currently in this test file...

ivanayov replied:
I did it with the idea that `_run_refresh()` creates new updater and the update functions are not repeated in `get_targetinfo()` which is practically the same as creating new updater and calling only `get_targetinfo()`. But I'm fine with changing it the other way.

ivanayov replied:
Changed without `_run_refresh()` for better readability.

sechkova reviewed (changes requested):
The test makes sense to me now, thanks for updating it. Some minor comments still but it looks ok to me.

jku reviewed (commented):
LGTM: @sechkova all yours to merge when you're happy

ivanayov commented:
Latest force-push is with addressing @sechkova 's [comment](https://github.com/theupdateframework/python-tuf/pull/1703#discussion_r768397735), nothing else has changed.

sechkova commented:
@jku  GH seems to believe you're still requesting changes and I cannot merge without your approval.

jku commented:
Sorry Ivana, I just made your life a bit harder by merging Teodoras test move.

Could you rebase?

(test looks really nice btw)

ivanayov commented:
@jku it's fine, I migrated it to `tests/test_updater_top_level_update.py`
